### PR TITLE
Support for `funcref`s, `ref.func`, and `table.grow` with `funcref`s

### PR DESCRIPTION
**Pull request by fitzgen**

### CODE REVIEW

File: crates/runtime/src/table.rs [hunk 1]
```diff
@@ -57,6 +57,17 @@ impl Table {
...
     }
 
+    /// Returns the type of the elements in this table.
+    pub fn element_type(&self) -> TableElementType {
+        match &*self.elements.borrow() {
+            TableElements::FuncRefs(_) => TableElementType::Func,
+            #[cfg(target_pointer_width = "32")]
+            TableElements::ExternRefs(_) => TableElementType::Val(ir::types::R32),
+            #[cfg(target_pointer_width = "64")]
+            TableElements::ExternRefs(_) => TableElementType::Val(ir::types::R64),
```
Comment (alexcrichton) [inline crates/runtime/src/table.rs:67]: FWIW this pattern with `#[cfg]` might be good to encapsulate somewhere with something like `ir::types::rsize()`

**Comment by fitzgen** [comment]
That definitely sounds like a better design, albeit more implementation effort. How strongly do you feel vs waiting until we actually care about the performance of `funcref`?

> We statically know all possible indices that can be referenced by `ref.func`,

I know that this is true at the spec level with the `refs` member of the validation context, but do you know if we actually retain this info anywhere? Not sure this is exposed by `wasmparser` or `cranelift_wasm`.

**Comment by alexcrichton** [comment]
I think it's fine to take the shortest path for now regardless of perf to a complete implementation of reference types, but looking forward to things like `table.copy` and `table.init` on `funcref` tables I feel like we're gonna be jumping through a lot of hoops trying to convert back and forth between `VMCallerCheckedAnyfunc` and `VMFuncRef`. I'm fine deferring to you, but it feels to me that it'll be basically equivalent the amount of work to take either strategy (in that I'm not sure one is that much more onerous than the other).

Currently AFAIK the `refs` member isn't exposed, so cranelift-wasm would likely need to construct this.

**Comment by fitzgen** [comment]
My one concern with your proposed design is just figuring out where to stick the new side tables for `VMCallerCheckedAnyfunc`s and how to plumb that through everywhere. Its the kind of thing that seems like it might run into the annoying `wasmtime`/`wasmtime-runtime` crate boundary again. I'll investigate a little more and see if I can't make it work.

**Comment by alexcrichton** [comment]
Perhaps we could temporarily allocate a `VMCallerCheckedAnyfunc` for every function in the module, and shove that into the `VMContext`? That way the index of the function is the same as the index into that table. We could later optimize that to not allocate such a big table for functions that aren't referenced.

**Comment by fitzgen** [comment]
Pushed a wip commit that switches to using `*const VMCallerCheckedAnyfunc` for `funcref`s. It compiles at least, but tests are failing.

**Comment by fitzgen** [comment]
@alexcrichton okay I think this is ready for review again.

File: crates/wasmtime/src/externals.rs [hunk 2]
```diff
@@ -23,7 +24,7 @@ use wasmtime_runtime::{self as runtime, InstanceHandle};
 #[derive(Clone)]
 pub enum Extern {
     /// A WebAssembly `func` which can be called.
-    Func(Func),
+    Func(Option<Func>),
```
Comment (alexcrichton) [inline crates/wasmtime/src/externals.rs:27]: The fallout from this change seems quite large (affecting the C API and many callers of the default API), but I can't figure out why this changed? 

While this makes sense for `Val` I don't think this makes sense for `Extern` which is acquired from an instance, and I don't think that instances can export a `null` funcref, right?

File: crates/environ/src/func_environ.rs [hunk 3]
```diff
@@ -604,9 +590,7 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
         });
 
         let element_size = match self.module.table_plans[index].style {
-            TableStyle::CallerChecksSignature => {
-                u64::from(self.offsets.size_of_vmcaller_checked_anyfunc())
-            }
+            TableStyle::CallerChecksSignature => self.pointer_type().bytes() as _,
```
Comment (alexcrichton) [inline crates/environ/src/func_environ.rs:593]: Could `u64::from` be used still to avoid `as`?

File: crates/runtime/src/instance.rs [hunk 4]
```diff
@@ -73,6 +75,9 @@ pub(crate) struct Instance {
     /// interrupted.
     pub(crate) interrupts: Arc<VMInterrupts>,
 
+    /// A canonical `ref.null func` value for this instance.
+    pub(crate) null_func_ref: Box<VMCallerCheckedAnyfunc>,
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs:79]: I saw somewhere else that this was store-local, but it's also here per-instance? Could this perhaps be `*mut` and stored in the `Store` instad?

File: crates/runtime/src/instance.rs [hunk 5]
```diff
@@ -1246,12 +1278,15 @@ fn initialize_tables(instance: &Instance) -> Result<(), InstantiationError> {
         }
 
         for (i, func_idx) in init.elements.iter().enumerate() {
-            let anyfunc = instance.get_caller_checked_anyfunc(*func_idx);
+            let anyfunc = instance.get_caller_checked_anyfunc(*func_idx).map_or(
+                instance.null_func_ref(),
+                |f: &VMCallerCheckedAnyfunc| {
+                    let p = f as *const VMCallerCheckedAnyfunc as *mut _;
+                    unsafe { NonNull::new_unchecked(p) }
```
Comment (alexcrichton) [inline crates/runtime/src/instance.rs:1285]: I think there's a safe conversion from `&T` to `NonNull<T>` to use here?

File: crates/wasmtime/src/func.rs [hunk 6]
```diff
@@ -157,6 +158,8 @@ macro_rules! getters {
             $($args: WasmTy,)*
             R: WasmTy,
         {
+            debug_assert!(unsafe { !self.export.anyfunc.as_ref().func_ptr.is_null() }, "b");
```
Comment (alexcrichton) [inline crates/wasmtime/src/func.rs:161]: `"b"`

File: crates/wasmtime/src/func.rs [hunk 7]
```diff
@@ -280,10 +284,11 @@ impl Func {
         });
         let (instance, export, trampoline) =
             crate::trampoline::generate_func_export(&ty, func, store).expect("generated func");
+        debug_assert!(unsafe { !export.anyfunc.as_ref().func_ptr.is_null() }, "c");
```
Comment (alexcrichton) [inline crates/wasmtime/src/func.rs:287]: `"c"`

File: crates/wasmtime/src/func.rs [hunk 8]
```diff
@@ -553,14 +567,21 @@ impl Func {
...
-        invoke_wasm_and_catch_traps(self.export.vmctx, &self.instance.store, || unsafe {
-            (self.trampoline)(
-                self.export.vmctx,
-                ptr::null_mut(),
-                self.export.address,
-                values_vec.as_mut_ptr(),
-            )
-        })?;
+        unsafe {
+            debug_assert!(!self.export.anyfunc.as_ref().func_ptr.is_null(), "d");
```
Comment (alexcrichton) [inline crates/wasmtime/src/func.rs:571]: This expression happens a lot in this block, so perhaps this could do this at the top?

```rust
let anyfunc = self.export.anyfunc.as_ref();
// ...
```

File: crates/wasmtime/src/func.rs [hunk 9]
```diff
@@ -578,22 +599,32 @@ impl Func {
...
+    ) -> NonNull<wasmtime_runtime::VMCallerCheckedAnyfunc> {
+        self.export.anyfunc
+    }
+
     pub(crate) fn from_wasmtime_function(
         export: wasmtime_runtime::ExportFunction,
         instance: StoreInstanceHandle,
-    ) -> Self {
+    ) -> Option<Self> {
+        if unsafe { export.anyfunc.as_ref().func_ptr.is_null() } {
```
Comment (alexcrichton) [inline crates/wasmtime/src/func.rs:612]: I think this answers my question about where the `None` came from, but I think that we're guaranteed `ExportFunction` is never a "null function", right?

File: crates/wasmtime/src/func.rs [hunk 10]
```diff
@@ -1189,6 +1220,7 @@ macro_rules! impl_into_func {
                     )
                     .expect("failed to generate export")
                 };
+                debug_assert!(unsafe { !export.anyfunc.as_ref().func_ptr.is_null() }, "a");
```
Comment (alexcrichton) [inline crates/wasmtime/src/func.rs:1223]: `"a"`

File: crates/wasmtime/src/instance.rs [hunk 11]
```diff
@@ -8,14 +8,15 @@ use wasmtime_jit::{CompiledModule, Resolver};
 use wasmtime_runtime::{InstantiationError, VMContext, VMFunctionBody};
 
 struct SimpleResolver<'a> {
+    store: Store,
```
Comment (alexcrichton) [inline crates/wasmtime/src/instance.rs:11]: Since there's already a lifetime here, perhaps this could be `&'a Store`?

File: crates/wasmtime/src/instance.rs [hunk 12]
```diff
@@ -50,8 +54,8 @@ fn instantiate(
             config.memory_creator.as_ref().map(|a| a as _),
             store.interrupts().clone(),
             host,
-            &*store.externref_activations_table() as *const _ as *mut _,
-            &*store.stack_map_registry() as *const _ as *mut _,
+            &**store.externref_activations_table() as *const _ as *mut _,
```
Comment (alexcrichton) [inline crates/wasmtime/src/instance.rs:57]: This got a little gnarly, mind annotating the `_` here?

File: crates/wasmtime/src/runtime.rs [hunk 13]
```diff
@@ -818,6 +819,7 @@ pub(crate) struct StoreInner {
     host_info: RefCell<HashMap<HostInfoKey, Rc<RefCell<dyn Any>>>>,
     externref_activations_table: Rc<VMExternRefActivationsTable>,
     stack_map_registry: Rc<StackMapRegistry>,
+    null_func_ref: Box<VMCallerCheckedAnyfunc>,
```
Comment (alexcrichton) [inline crates/wasmtime/src/runtime.rs:822]: Ah it looks like there is indeed a null funcref per-store, but perhaps this could be reused for each `Instance`?

File: crates/wasmtime/src/types.rs [hunk 14]
```diff
@@ -106,25 +106,7 @@ impl ValType {
...
-            ir::types::I32 => Some(ValType::I32),
-            ir::types::I64 => Some(ValType::I64),
-            ir::types::F32 => Some(ValType::F32),
-            ir::types::F64 => Some(ValType::F64),
-            ir::types::I8X16 => Some(ValType::V128),
-            #[cfg(target_pointer_width = "64")]
-            ir::types::R64 => Some(ValType::ExternRef),
-            #[cfg(target_pointer_width = "32")]
-            ir::types::R32 => Some(ValType::ExternRef),
+            ValType::ExternRef => Some(wasmtime_runtime::ref_type()),
```
Comment (alexcrichton) [inline crates/wasmtime/src/types.rs:109]: Could this perhaps go through the `Store` instead? One day if we ever support cross-compilation this'd be through that instead of `wasmtime-runtime`

File: crates/wasmtime/src/values.rs [hunk 15]
```diff
@@ -94,7 +101,13 @@ impl Val {
                     .insert_with_gc(x.inner, store.stack_map_registry());
                 ptr::write(p as *mut *mut u8, externref_ptr)
             }
-            _ => unimplemented!("Val::write_value_to"),
+            Val::FuncRef(None) => ptr::write(p, 0),
```
Comment (alexcrichton) [inline crates/wasmtime/src/values.rs:104]: I think this should write the store "null funcref" singleton instead of a null pointer, right?

File: crates/wasmtime/src/values.rs [hunk 16]
```diff
@@ -116,7 +129,14 @@ impl Val {
                     }))
                 }
             }
-            _ => unimplemented!("Val::read_value_from: {:?}", ty),
+            ValType::FuncRef => {
+                let func = ptr::read(p as *const *mut runtime::VMCallerCheckedAnyfunc);
+                if let Some(f) = NonNull::new(func) {
```
Comment (alexcrichton) [inline crates/wasmtime/src/values.rs:134]: Is `NonNull::new` here necessary?  This is related to my question above too I think.

I figured that `ref.null` would be implemented as returning the null singleton rather than a literal null pointer. (to make `table.set` a simple store instead of having a branch to store a non-null value)

File: crates/wasmtime/src/values.rs [hunk 17]
```diff
@@ -211,51 +232,47 @@ impl From<Option<ExternRef>> for Val {
...
-        Val::FuncRef(f) => {
-            let f = f.wasmtime_function();
-            wasmtime_runtime::VMCallerCheckedAnyfunc {
-                func_ptr: f.address,
-                type_index: f.signature,
-                vmctx: f.vmctx,
-            }
-        }
+        Val::FuncRef(None) => store.null_func_ref(),
+        Val::FuncRef(Some(f)) => f.caller_checked_anyfunc(),
```
Comment (alexcrichton) [inline crates/wasmtime/src/values.rs:256]: An idle thought, but we could perhaps have something like:

```rust
impl Func {
    pub(crate) fn into_raw(self, dst: &Store) -> Result<NonNull<VMCallerCheckedAnyfunc>> {
        // ... 
    }
}
```

to bake in how you can't access the raw value without using the right store. Anyway not necessary to do here, just a thought!

File: crates/wasmtime/src/values.rs [hunk 18]
```diff
@@ -211,51 +232,47 @@ impl From<Option<ExternRef>> for Val {
...
 
-pub(crate) fn from_checked_anyfunc(
-    item: wasmtime_runtime::VMCallerCheckedAnyfunc,
+pub(crate) unsafe fn from_checked_anyfunc(
+    anyfunc: NonNull<wasmtime_runtime::VMCallerCheckedAnyfunc>,
     store: &Store,
 ) -> Val {
-    if item.type_index == wasmtime_runtime::VMSharedSignatureIndex::default() {
-        return Val::ExternRef(None);
+    if anyfunc.as_ref().func_ptr.is_null() {
```
Comment (alexcrichton) [inline crates/wasmtime/src/values.rs:265]: Another passing though, perhaps `fn is_null(&self) -> bool` could be added to `VMCallerCheckedAnyfunc`

File: crates/wast/src/spectest.rs [hunk 19]
```diff
@@ -35,7 +35,7 @@ pub fn link_spectest(linker: &mut Linker) -> Result<()> {
     linker.define("spectest", "global_f64", g)?;
 
     let ty = TableType::new(ValType::FuncRef, Limits::new(10, Some(20)));
-    let table = Table::new(linker.store(), ty, Val::ExternRef(None))?;
+    let table = Table::new(linker.store(), ty, Val::FuncRef(None))?;
```
Comment (alexcrichton) [inline crates/wast/src/spectest.rs:38]: Nice 👍

File: crates/wasmtime/src/types.rs (hunk 14 above)
Comment (fitzgen) [reply crates/wasmtime/src/types.rs:109]: `wasmtime-environ` is the thing that should handle the cross compilation as it is about the environment of the compilation target, while `wasmtime-runtime` is always compiled into the target. So if we put it in `wasmtime-runtime`, then it is always going to reflect the compilation target.

If we wanted to support cross compilation at the `wasmtime` API level, then we would have to stop linking `wasmtime-runtime` directly, which would force us to (correctly) revisit calls to this function from the `wasmtime` crate. So I think we should be good here.

File: crates/wasmtime/src/externals.rs (hunk 2 above)
Comment (fitzgen) [reply crates/wasmtime/src/externals.rs:27]: I... *think* you are correct. I'll look into it.

File: crates/runtime/src/instance.rs (hunk 5 above)
Comment (fitzgen) [reply crates/runtime/src/instance.rs:1285]: I didn't realize that `NonNull<T>` implements `From<&T>`!

File: crates/runtime/src/instance.rs (hunk 4 above)
Comment (fitzgen) [reply crates/runtime/src/instance.rs:79]: FWIW, we don't use pointer equality for `VMCallerCheckedAnyfunc`s, so this would just be a memory optimization that saves two words per instance.

Comment (alexcrichton) [reply crates/runtime/src/instance.rs:79]: Ah yeah that's all I'm imagining, fewer small allocations here-and-there. It also sort of helps the case of "I precompiled my wasm" since we'd presumably statically allocate this anyfunc instead of trying to dynamically allocate it.

File: crates/wasmtime/src/types.rs (hunk 14 above)
Comment (alexcrichton) [reply crates/wasmtime/src/types.rs:109]: It's true yeah we're a *long* way away from wasmtime supporting AOT compilation, but I just figured this'd be a relatively easy future-proofing.

Comment (fitzgen) [reply crates/wasmtime/src/types.rs:109]: I think we are as future proofed as it makes sense to be given that this function is safely inside `wasmtime-runtime`, as mentioned above.

I think our best path forward for AOT cross compilation is a `wasmtime-aot` crate.

**Comment by fitzgen** [comment]
@alexcrichton okay, I've deduped the null func refs into the store, and made `Extern::Func` non-nullable.

Still need to write a couple tests that pass in `Val::Func(None)` to various things, but I think you can take another look a this while I do that.

**Comment by alexcrichton** [comment]
Looks great to me! I've got one final question though before landing, which I just thought of after reading this. Could we be even *more* clever and make the null funcref a literal null pointer? It looks like you were worried about the cost of `call_indirect` in terms of number of checks, but I think that it's the same number of checks as before, we're just testing `anyfunc_ptr` for whether it's null instead of `func_addr`, which we assume is always non-null.

With that in place translation of `ref.null` would be super simple, just a bag of zeros all the time!

**Comment by fitzgen** [comment]
> Looks great to me! I've got one final question though before landing, which I just thought of after reading this. Could we be even _more_ clever and make the null funcref a literal null pointer? It looks like you were worried about the cost of `call_indirect` in terms of number of checks, but I think that it's the same number of checks as before, we're just testing `anyfunc_ptr` for whether it's null instead of `func_addr`, which we assume is always non-null.
> 
> With that in place translation of `ref.null` would be super simple, just a bag of zeros all the time!

The concern wasn't about the path for `call_indirect` with a null funcref (which should be a super rare event, and leads to a trap anyways, so it isn't ever gonna be fast). The concern was introducing an extra potential cache miss in front of every non-trapping `call_indirect` before we actually end up calling the function.

Originally, we couldn't use a null pointer for a null func ref, but after all the refactorings this PR has been through, that might not be the case anymore. Let me double check. Either way, I think `ref.null` will be cheap enough either way. But this actually made me discover a latent bug: `cranelift-wasm` doesn't allow you customize `ref.null`'s translation (currently). It also always assumes that all references are of the reference type (which is not true for `funcref`s for us). So I have a bit more work to do here either way.

**Comment by alexcrichton** [comment]
Oh I'm not actually worried about the cost of `ref.null`, it's more that using a null pointer for `ref.null` felt clearer and easier to understand.

Could you elaborate on the cache misses, though? With this current incantation a `call_indirect` is: load anyref, load funcptr, check null, load type, check match, call. With a null pointer representation it would instead be load anyref, check null, load type, check match, load funcptr, call. I'm not sure how the cache misses would be different there?

**Comment by fitzgen** [comment]
The extra potential cache miss is from changing table elements from `VMCallerCheckedAnyfunc` to a pointer to a `VMCallerCheckedAnyfunc` (regardless whether that is a nullable pointer or not).

It makes it so that there is an extra pointer chase to get to the `func_ptr` member of an element in the table.

**Comment by alexcrichton** [comment]
Ah ok, I remember talking about that yeah. I thought you meant that the extra cache miss was about the representation of the null function in the sense of is it a null pointer or is it a sentinel.

**Comment by fitzgen** [comment]
phew. okay.

* I made `ref.null func` be a null pointer, rather than a `VMCallerCheckedAnyfunc` with a null `func_ptr` member.
* similarly, `VMCallerCheckedAnyfunc::func_ptr` is now `NonNull`.
* I added the extra `translate_*` hooks to `cranelift_wasm::FuncEnvironment` to allow for non-uniform reference types representations.

@alexcrichton I think this is ready for another (and hopefully the last!) round of review. Thanks for your patience with this one!

File: crates/wasmtime/src/values.rs [hunk 20]
```diff
@@ -211,51 +229,46 @@ impl From<Option<ExternRef>> for Val {
...
+        debug_assert!(
+            anyfunc.as_ref().type_index != wasmtime_runtime::VMSharedSignatureIndex::default()
+        );
+        let instance_handle = wasmtime_runtime::InstanceHandle::from_vmctx(anyfunc.as_ref().vmctx);
+        let export = wasmtime_runtime::ExportFunction { anyfunc };
+        let instance = store.existing_instance_handle(instance_handle);
+        let f = Func::from_wasmtime_function(export, instance);
+        Val::FuncRef(Some(f))
+    } else {
+        return Val::FuncRef(None);
```
Comment (alexcrichton) [inline crates/wasmtime/src/values.rs:272]: I think the `return` here can go away, or this may also benefit from a early-return `match`:

```rust
let anyfunc = match NonNull:new(anyfunc) {
    Some(f) => f,
    None => return Val::FuncRef(None),
};
// ...
